File: implementations/java/vanilla/src/main/java/org/cwi/waebric/checker/ModuleChecker.java
package org.cwi.waebric.checker;

import java.io.File;
import java.util.List;

import org.cwi.waebric.checker.exception.NonExistingModuleException;
import org.cwi.waebric.checker.exception.SemanticException;
import org.cwi.waebric.parser.ast.DefaultNodeVisitor;
import org.cwi.waebric.parser.ast.module.Import;
import org.cwi.waebric.parser.ast.module.Module;
import org.cwi.waebric.parser.ast.module.ModuleId;
import org.cwi.waebric.util.ModuleRegister;

public class ModuleChecker extends DefaultNodeVisitor {

	/**
	 * Exceptions
	 */
	private final List<SemanticException> exceptions;

	/**
	 * Construct checker
	 * @param exceptions
	 */
	public ModuleChecker(List<SemanticException> exceptions) {
		this.exceptions = exceptions;
	}
	
	@Override
	public void visit(Module module) {
		module.getIdentifier().accept(this);
		for(Import imprt: module.getImports()) {
			imprt.accept(this);
		}
	}
	
	@Override
	public void visit(ModuleId identifier) {
		String path = ModuleRegister.getPath(identifier);
		File file = new File(path);
		if(! file.isFile()) {
			exceptions.add(new NonExistingModuleException(identifier));
		}
	}
	
}

File: implementations/csharp/vanilla/Checker/ModuleChecker.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;
using Parser.Ast.Statements;
using Parser.Ast.Module;
using Common;
using Checker.Exceptions;
using Parser.Ast;


namespace Checker
{
    /// <summary>
    /// Checker which checks module references
    /// </summary>
    public class ModuleChecker : SyntaxNodeVisitor
    {
        #region Private Members

        private List<Exception> ExceptionList;

        #endregion

        #region Public Methods

        public ModuleChecker(List<Exception> exceptionList)
        {
            ExceptionList = exceptionList;
        }

        public new void Visit(Module module)
        {
            //Check filename
            module.GetModuleId().AcceptVisitor(this);

            //Get dependencies and check their existence
            List<Module> DependencyList = ModuleCache.Instance.GetDependencies(module);
            foreach (Module moduleDependency in DependencyList)
            {
                foreach (Import import in moduleDependency.GetImports())
                {   //Check imports of dependency's
                    import.AcceptVisitor(this);
                }
            }
        }

        public new void Visit(Import import)
        {
            //Check if import exists
            import.GetModuleId().AcceptVisitor(this);
        }

        public new void Visit(ModuleId moduleId)
        {
            //Check if path exists
            String path = ModuleCache.Instance.GetPath(moduleId);
            if (!File.Exists(path))
            {
                ExceptionList.Add(new NonExistingModule(moduleId));
            }
        }

        #endregion
    }
}
